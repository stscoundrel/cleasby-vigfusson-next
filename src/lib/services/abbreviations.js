import { findAbbreviations, findWorksAndAuthors } from 'cleasby-vigfusson-abbreviations'
import { abbreviate } from 'abbreviatrix'

/**
 * Combine abbreviations from definitions array.
 * Accepts finder method to use.
 */
const combineAbbreviations = ({ definitions }, find) => {
  const combinedAbbreviations = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = find(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

export const getAbbreviations = (entry) => ({
  common: combineAbbreviations(entry, findAbbreviations),
  works: combineAbbreviations(entry, findWorksAndAuthors),
})

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (content, abbreviations) => {
  const combinedAbbrs = [...abbreviations.common, ...abbreviations.works]
  let result = content

  combinedAbbrs.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export default {
  getAbbreviations,
  addAbbreviationsToContent,
}
