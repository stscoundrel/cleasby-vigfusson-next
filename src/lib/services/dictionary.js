const { getDictionary } = require('cleasby-vigfusson-dictionary')
const { VALID_AS_FIRST } = require('old-norse-alphabet')
const { slugifyWord, slugifyLetter } = require('../utils/slugs')

let cachedDictionary = null

const addSlugs = (words) => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = () => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  cachedDictionary = formattedWords

  return formattedWords
}

export const getByLetter = (letter) => {
  const words = getAllWords()
  const byLetter = words.filter((entry) => (
    entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))

  return byLetter
}

export const getWord = (word) => getAllWords().filter((entry) => entry.slug === word)[0]

export const getAlphabet = () => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getAlphabet,
}
