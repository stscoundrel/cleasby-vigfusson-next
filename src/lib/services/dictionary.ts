import { getDictionary, DictionaryEntry as RawDictionaryEntry } from 'cleasby-vigfusson-dictionary'
import { VALID_AS_FIRST } from 'old-norse-alphabet'
import { oldNorseSort } from 'old-norse-alphabet-sort'
import { slugifyWord, slugifyLetter } from '../utils/slugs'

export interface DictionaryEntry extends RawDictionaryEntry {
  slug: string,
}

export interface DictionaryEntryDTO {
  word: string,
  slug: string
}

export interface AlphabetLetter {
  letter: string,
  slug: string
}

let cachedDictionary: DictionaryEntry[] | null = null

const addSlugs = (words: RawDictionaryEntry[]): DictionaryEntry[] => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = (): DictionaryEntry[] => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  /**
   * Ensure headwords are lowercased.
   * Original source has some in all caps.
   * That is true to the layout of the book,
   * but makes little sense in website.
   */
  const dictionary = formattedWords.map((entry) => ({ ...entry, word: entry.word.toLowerCase() }))

  cachedDictionary = dictionary

  return dictionary
}

export const getByLetter = (letter: string): DictionaryEntryDTO[] => {
  const words = getAllWords()
  const byLetter = words
    .filter((entry) => (
      entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))
    .map((entry) => {
      // Simpler DTO dictionary entry.
      const { word, slug } = entry
      return { word, slug }
    })
    .sort((a, b) => oldNorseSort(a.word, b.word))

  return byLetter
}

export const getWord = (slug: string): DictionaryEntry => (
  getAllWords().filter((entry) => entry.slug === slug)[0]
)

export const getRandomEntries = (): DictionaryEntry[] => (
  // Return entries fit to be randomized "teasers"
  // Therefore, content should be short, but not too short.
  getAllWords()
    .sort(() => Math.random() - 0.5)
    .filter((entry) => entry.definitions[0].length < 50 && entry.definitions[0].length > 15)
    .slice(0, 36)
    .sort((a, b) => oldNorseSort(a.word, b.word))
)

export const getSimilarWords = (entry: DictionaryEntry) : DictionaryEntry[] => getAllWords()
  .filter((dEntry) => dEntry.word.toLocaleLowerCase() === entry.word.toLocaleLowerCase()
  && dEntry.slug !== entry.slug)

export const getAlphabet = (): AlphabetLetter[] => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getRandomEntries,
  getAlphabet,
}
