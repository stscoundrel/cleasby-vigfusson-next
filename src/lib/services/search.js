import { markWords } from 'markari'

const formatResults = (results, search, criteria) => {
  const formattedResults = results.map((result) => {
    const foundIn = []

    if (criteria.includes('definitions')) {
      result.definitions.forEach((definition) => {
        if (definition.toLowerCase().includes(search.toLowerCase())) {
          const highlighterDefinition = markWords(search, definition)
          foundIn.push(highlighterDefinition)
        }
      })
    }

    if (foundIn.length === 0) {
      const highlightedHeadword = markWords(search, result.word)
      foundIn.push(`In headword: ${highlightedHeadword}`)
    }

    return {
      ...result,
      foundIn,
    }
  })

  return formattedResults
}

export const searchDictionary = (search, dictionary, criteria = ['headword', 'definitions']) => {
  const filteredSearch = search.toLowerCase()

  const results = dictionary.filter((entry) => {
    let matchesSearch = false

    if (criteria.includes('headword')) {
      if (entry.word.toLowerCase().includes(filteredSearch)) {
        matchesSearch = true
      }

      if (entry.slug.includes(filteredSearch)) {
        matchesSearch = true
      }
    }

    if (criteria.includes('definitions')) {
      entry.definitions.forEach((definition) => {
        if (definition.toLowerCase().includes(filteredSearch)) {
          matchesSearch = true
        }
      })
    }

    return matchesSearch
  })

  const formattedResult = formatResults(results, search, criteria)

  return formattedResult
}

export default {
  searchDictionary,
}
