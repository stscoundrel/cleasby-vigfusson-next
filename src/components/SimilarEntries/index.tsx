import WordLink from 'components/WordLink'
import { DictionaryEntry } from 'lib/services/dictionary'
import { capitalize } from 'lib/utils/strings';

interface SimilarEntriesProps {
  entries: DictionaryEntry[]
}

const postfixMap = new Map()
postfixMap.set('1', 'I')
postfixMap.set('2', 'II')
postfixMap.set('3', 'III')
postfixMap.set('4', 'IV')
postfixMap.set('5', 'V')
postfixMap.set('6', 'VI')
postfixMap.set('7', 'VII')
postfixMap.set('8', 'VIII')
postfixMap.set('9', 'IX')
postfixMap.set('10', 'X')

export default function SimilarEntries({ entries }: SimilarEntriesProps) {
  const parsePostfix = (entryToParse: DictionaryEntry): string => {
    let postfix = 'I';
    const lastChar = entryToParse.slug.slice(-1)

    if (postfixMap.has(lastChar)) {
      postfix = postfixMap.get(lastChar)
    }

    return `${capitalize(entryToParse.word)} (${postfix})`
  }

  if (entries.length === 0) {
    return null
  }

  return (
    <>
      <h4>Similar entries:</h4>
      <ul>
        {entries.map((entry, index) => (
          <li key={`similar-link-${index}`}>
            <WordLink data={{ ...entry, word: parsePostfix(entry) }} useLowerCase={false} />
          </li>
        ))}
      </ul>
    </>
  )
}
