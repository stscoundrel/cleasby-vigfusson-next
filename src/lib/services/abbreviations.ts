import abbreviationsService from 'cleasby-vigfusson-abbreviations'
import { abbreviate } from 'abbreviatrix'
import { DictionaryEntry } from 'lib/services/dictionary'

const { findAbbreviations, findWorksAndAuthors, getWorksAndAuthorsMapping } = abbreviationsService

export interface Abbreviation{
  abbreviation: string,
  explanation: string
}

export interface CombinedAbbreviations {
  common: Abbreviation[],
  works: Abbreviation[],
}

/**
 * Combine abbreviations from definitions array.
 * Accepts finder method to use.
 */
const combineAbbreviations = (
  { definitions }: DictionaryEntry,
  abbreviationFinder: (s: string) => Map<string, string>,
): Abbreviation[] => {
  const combinedAbbreviations: Abbreviation[] = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = abbreviationFinder(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

export const getAbbreviations = (entry: DictionaryEntry): CombinedAbbreviations => ({
  common: combineAbbreviations(entry, findAbbreviations),
  works: combineAbbreviations(entry, findWorksAndAuthors),
})

export const getAllSorces = (): Abbreviation[] => {
  const abbrs = getWorksAndAuthorsMapping()
  return Array
    .from(abbrs, ([abbreviation, explanation]) => ({ explanation, abbreviation }))
    .sort((a, b) => a.abbreviation.localeCompare(b.abbreviation))
}

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (
  content: string,
  abbreviations: CombinedAbbreviations,
): string => {
  const combinedAbbrs = [...abbreviations.common, ...abbreviations.works]
  let result = content

  combinedAbbrs.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export default {
  getAbbreviations,
  addAbbreviationsToContent,
}
