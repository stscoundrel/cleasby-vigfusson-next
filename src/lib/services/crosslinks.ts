import { getCrosslinks as getAllCrossLinks, Crosslink, getOldNorseCrosslinks } from 'scandinavian-dictionary-crosslinker'
import { DictionaryEntry } from 'lib/services/dictionary'

export type { Crosslink } from 'scandinavian-dictionary-crosslinker'

let crossLinkCache: Record<string, Crosslink[]> | null = null

const populateCrossLinks = () => {
  if (!crossLinkCache) {
    crossLinkCache = getAllCrossLinks()
  }

  return crossLinkCache
}

export const getCrossLinks = (entry: DictionaryEntry): Crosslink[] => {
  const crosslinks = populateCrossLinks()
  if (Object.prototype.hasOwnProperty.call(crosslinks, entry.slug)) {
    return getOldNorseCrosslinks(entry.slug)
  }

  return []
}
