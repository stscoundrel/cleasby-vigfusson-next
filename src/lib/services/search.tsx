import { markWords } from 'markari'
import { getAbbreviations, addAbbreviationsToContent } from 'lib/services/abbreviations'
import { DictionaryEntry } from './dictionary'

export interface SearchResult extends DictionaryEntry{
  foundIn: string[],
}

export type Criteria = 'all' | 'headword' | 'definitions'

const formatResults = (results, search, criteria): SearchResult[] => {
  const formattedResults = results.map((result) => {
    const foundIn: string[] = []

    if (criteria.includes('definitions')) {
      result.definitions.forEach((definition) => {
        if (definition.toLowerCase().includes(search.toLowerCase())) {
          const abbreviations = getAbbreviations(result)
          const highlighterDefinition = markWords(search, definition)
          const abbrTaggedDefinition = addAbbreviationsToContent(
            highlighterDefinition,
            abbreviations,
          )
          foundIn.push(abbrTaggedDefinition)
        }
      })
    }

    if (foundIn.length === 0) {
      const highlightedHeadword = markWords(search, result.word)
      foundIn.push(`In headword: ${highlightedHeadword}`)
    }

    return {
      ...result,
      foundIn,
    }
  })

  return formattedResults
}

export const searchDictionary = (
  search: string,
  dictionary: DictionaryEntry[],
  criteria: Criteria[] = ['headword', 'definitions'],
): SearchResult[] => {
  const filteredSearch = search.toLowerCase()

  const results = dictionary.filter((entry) => {
    let matchesSearch = false

    if (criteria.includes('headword')) {
      if (entry.word.toLowerCase().includes(filteredSearch)) {
        matchesSearch = true
      }

      if (entry.slug.includes(filteredSearch)) {
        matchesSearch = true
      }
    }

    if (criteria.includes('definitions')) {
      entry.definitions.forEach((definition) => {
        if (definition.toLowerCase().includes(filteredSearch)) {
          matchesSearch = true
        }
      })
    }

    return matchesSearch
  })

  const formattedResult = formatResults(results, search, criteria)

  return formattedResult
}

export default {
  searchDictionary,
}
