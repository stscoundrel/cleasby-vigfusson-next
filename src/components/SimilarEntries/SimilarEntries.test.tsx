import renderer from 'react-test-renderer'
import { DictionaryEntry } from 'lib/services/dictionary'
import SimilarEntries from './index'

describe('Similar entries component', () => {
  const entries: DictionaryEntry[] = [
    {
      word: 'fadhurlikhet',
      partOfSpeech: [
        'nn',
      ],
      grammaticalAspect: '',
      information: '',
      definitions: [
        '1)  faderligt sinne. til mere kerlex ok faderlighetz jertekn  Lg 3: 562 . ',
        '2)  faderlighet.  " ss titel. som edher fadher lighet vel fortencker " FM 271 (1506, bref från Svante Sture till biskop Lauentius i Åbo). bidhiom vi thina fadhirlikhet  Gr 320 .',
      ],
      alternativeForms: [
        'fadhirlikhet . ',
        'fadherlighet . faderlighet),',
      ],
      slug: 'fadhurlikhet-2',
    },
  ]

  test('Matches the snapshot', () => {
    const tree = renderer.create(
      <SimilarEntries entries={entries} />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
