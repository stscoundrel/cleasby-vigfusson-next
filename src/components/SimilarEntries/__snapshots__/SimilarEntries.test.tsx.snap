// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Similar entries component Matches the snapshot 1`] = `
[
  <h4>
    Similar entries:
  </h4>,
  <ul>
    <li>
      <a
        className="link"
        href="/word/fadhurlikhet-2"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Fadhurlikhet (II)
      </a>
    </li>
  </ul>,
]
`;
